### PR TITLE
Create useragent if one does not exist

### PR DESCRIPTION
**Pull request by Lash-L**
Seemingly, the useragent is never created unless there is an exception with a specific error code(And users are getting a 403 error code, not a 400 error code). In my opinion, it is more logical to just create the useragent once at the beginning and still reset it later when needed. That way you also aren't starting with a failed attempt, you can immediately start with a successful attempt. 

Users are seemingly having issues with a request with an empty useragent being rejected.

### CODE REVIEW

**Comment by drothenberger** [comment]
The MyQ service is now returning a 403 error instead of a 400 error. Wouldn't it make more sense to also trigger get_useragent() on a 403 errors instead of just setting the user agent initially?

E.g.,
``` python

                if err.status == 403 and attempt == 0:
                    _LOGGER.debug(
                        "Received error status 403, bad request. Will refresh user agent."
                    )
                    await self._get_useragent()
```

**Comment by Lash-L** [comment]
Personally I don't think it makes sense to send a request that you know will fail - I think adding a check for err.status == 403 or err.status == 400 would be helpful though. But logically to me, you should start with a user agent

I've added 403 as a check

**Comment by drothenberger** [comment]
> Personally I don't think it makes sense to send a request that you know will fail - I think adding a check for err.status == 403 or err.status == 400 would be helpful though. But logically to me, you should start with a user agent
> 
> I've added 403 as a check

I get the impression that at some point in the past, MyQ was blocking requests with certain user agents, so the library was changed to try to avoid setting it at all unless it was necessary. It seems that MyQ is now requiring the User Agent, but if that should change again in the future, setting the User Agent proactively may cause problems again.

**Review by drothenberger — COMMENTED** [review]
The error message should reflect the error code received, instead of always logging that 400 was returned.

**Review by mikeodr — APPROVED** [review]

**Comment by Lash-L** [comment]
> I get the impression that at some point in the past, MyQ was blocking requests with certain user agents, so the library was changed to try to avoid setting it at all unless it was necessary. It seems that MyQ is now requiring the User Agent, but if that should change again in the future, setting the User Agent proactively may cause problems again.

Fair enough I suppose. I don't have a myq - so I have no skin in the game here, I just saw the issue on HA, saw people realized the problem, but no one had done a PR yet so I figured I'd help out and get one out.

If you'd like to open a PR as an alternative without setting the user_agent, I'm more than happy to close this one. If you want to wait for a maintainer to give their feedback, that is completely find by me as well. Like I said, I have no skin in the game, so I'm rather indifferent

But it doesn't fully make sense to me to make a change that you know will still fail at least one time to try to prevent a potential error if they rework myq again

**Comment by drothenberger** [comment]
> If you'd like to open a PR as an alternative without setting the user_agent, I'm more than happy to close this one. If you want to wait for a maintainer to give their feedback, that is completely find by me as well. Like I said, I have no skin in the game, so I'm rather indifferent

I'll let the maintainer decide. It would be easy enough merge the PR without the initial setting of User Agent, if the maintainer thinks that is the best approach.

Thanks for making the PR, even though the issue is not affecting you!

**Comment by Lash-L** [comment]
@ehendrix23 Since you are seemingly the most recent active contributor to this repository and you are the codeowner on HA, I figured I'd tag you to make you aware of this PR

**Comment by drothenberger** [comment]
I confirm this PR resolves the issue in HA. I downloaded the ZIP from this PR and put it into my config directory, then restarted HA. The error went away.

**Review by bdraco — APPROVED** [review]
Confirmed fix 👍

**Comment by bdraco** [comment]
Looks good. I don't have merge rights to this repo so we need to wait for ehendrix23 to be available to merge and release.

**Comment by firstof9** [comment]
Might be a good idea to find the UserAgent the app uses and just set it to that?

**Comment by bdraco** [comment]
> Might be a good idea to find the UserAgent the app uses and just set it to that?

That strategy was tried in the past but didn't work out

**Comment by kj4qyf** [comment]
What is a user agent and how do I create it and use in HA ?

**Comment by drothenberger** [comment]
> What is a user agent and how do I create it and use in HA ?

User-Agent is a header in the HTTP requests made by this library to the MyQ servers.

If you want to get the fix into HA before the next HA release, see [these instructions](https://github.com/home-assistant/core/issues/99947#issuecomment-1712572456).

**Review by jhansche — APPROVED** [review]

**Review by ChrisCarini — APPROVED** [review]
LGTM! Let's get this released.

**Comment by starsoccer** [comment]
While I am not against this change, it seems like this isnt the root cause of the issue. See below threads. Ive tried to apply this fix manually myself and it does not resolve the issue for me. I still get 403 errors.

https://github.com/hjdhjd/myq/issues/411#issuecomment-1712971457
https://github.com/home-assistant/core/issues/99947

**Comment by hjdhjd** [comment]
I would add: https://github.com/home-assistant/core/issues/99947#issuecomment-1713940465

As a further explanation for why adding a UA is not an advisable solution to this and may ultimately create another unnecessary game of cat-and-mouse with the API in the open source myQ community.

I would encourage @ehendrix23 and @ChrisCarini (and the rest of the developers) to rethink this approach, particularly given the history of this API over the years.

**Comment by joshtbernstein** [comment]
If it's not a UA issue, which it seems to not 100% be, would it be worth building in multiple URLs to try if the load balanced ones fail?

For example, if `devices.myq-cloud.com`/`account-devices-gdo.myq-cloud.com` fail, try `devices-east.myq-cloud.com`/`account-devices-gdo-east.myq-cloud.com` and/or `devices-west.myq-cloud.com`/`account-devices-gdo-west.myq-cloud.com`.

**Comment by starsoccer** [comment]
> If it's not a UA issue, which it seems to not 100% be, would it be worth building in multiple URLs to try if the load balanced ones fail?
> 
> For example, if `devices.myq-cloud.com`/`account-devices-gdo.myq-cloud.com` fail, try `devices-east.myq-cloud.com`/`account-devices-gdo-east.myq-cloud.com` and/or `devices-west.myq-cloud.com`/`account-devices-gdo-west.myq-cloud.com`.

That may be a better solution. The main question I think is does this actually fix the issue for everyone. I can only comment for myself, but trying to set to east only didnt seem to fix it. I didnt try to change it to west, but Id be surprised if that works. My myq has been more or less totally offline since the issue started with no change resolving it

**Comment by 86lava** [comment]
None of the earlier posts appear to acknowledge that NOT EVERYONE has Python 3.11 INSTALLED. 

I have been following this issue since shortly after it surfaced.  Shortly after @lygris, posted his user_agent resolution, I set out to correct my install with success.  When I couldn't find the Python3.11 directory, learned that my setup was running Python 3.9.9 - so I deviated from the instructions accordingly and everything worked.  
  
When the zip file "fix" surfaced a few hours later, I downloaded and opened the package, noting that this zip file method can only work when the setup is running Python 3.11.  My [comment a few days ago in the core#99947](https://github.com/home-assistant/core/issues/99947#issuecomment-1712581193) discussion also seemed to have gone unnoticed.

**Comment by lygris** [comment]
@hjdhjd Honestly curious where or how you found the info about it being an azure load balancer issue? Also on the user agent not being required portion, I think it is safe to say that is false. Nobody is saying it's a required HTTP header, but it seems fairly clear myq/chamberlain has implemented, knowingly or not, some form of UA blocking. Whether they put in a new load balancer or turned on a feature, that unbeknownst to them, blocks blank UA's or they willingly are blocking blank UA's doesn't really matter. 

@starsoccer What region of the world are you in? Overall it seems like the fix worked for most people. There were a few that it didn't but i'm leaning towards them not putting the code change in correctly. My only other suspicion is regionally there are differences in how the myq cloud is setup.

**Comment by starsoccer** [comment]
@lygris I am East coast and changed my code to point at east which didnt work. I can only speak to my setup personally, but no change has fixed the issue for my since roughly when this started for everyone else around the .9 update.

I am happy to apply some other changes if anyone has some to test and see if it resolve it for me

**Comment by lygris** [comment]
I upgraded HA core today and of course myq broke, so I redid the fix and set my UA to "HomeAssistant" and rebooted. received a 403 error on startup. Then went and changed my UA to a well known and widely used UA, rebooted, and myq started working. This is too small of a sample size to fully confirm it's a UA issue, but i think it's fairly telling. Others in HA core issue thread also did state they changed their UA's per my instructions to something like "hass" or other HA related terms/names and reported it did not work. Some however made up a seemingly random string and it did work.

@starsoccer try this if your request.py is back to the default
`docker exec -it homeassistant sed -i.bak "34s/self._useragent = None/self._useragent = str(\"AppleWebKit\/537.36\")/" /usr/local/lib/python3.11/site-packages/pymyq/request.py`

**Comment by hjdhjd** [comment]
> @hjdhjd Honestly curious where or how you found the info about it being an azure load balancer issue? Also on the user agent not being required portion, I think it is safe to say that is false. Nobody is saying it's a required HTTP header, but it seems fairly clear myq/chamberlain has implemented, knowingly or not, some form of UA blocking. Whether they put in a new load balancer or turned on a feature, that unbeknownst to them, blocks blank UA's or they willingly are blocking blank UA's doesn't really matter.
> 
> @starsoccer What region of the world are you in? Overall it seems like the fix worked for most people. There were a few that it didn't but i'm leaning towards them not putting the code change in correctly. My only other suspicion is regionally there are differences in how the myq cloud is setup.

I'm intentionally setting aside the first part of your question - I'm really not interested in a public dialog on the topic beyond pointing folks to take the time to understand/investigate/research the myQ API, how Azure regions work, and how Azure load balancing (when configured) work. A few DNS queries and a couple of web searches and you can put the picture together...it's not complex. 😉

The second: **no, the evidence does not indicate that myQ has instituted UA blocking.** Homebridge myQ - my plugin - and the API library I wrote that it uses (which pymyq uses as a reference for it's own implementation), does not use UA, and will not be unless there is an absolute need to do so. I really wish people would stop making statements that aren't supported by a significant body of evidence. There may be some users who are having issues where setting a UA seems to mitigate the issue for the time being - but that is not the same as stating "it seems clear that myQ has implemented...some form of UA blocking". They have not as far as I can tell. **If they had, then every implementation, including my own, would be broken. That is simply not the case.** I use the myQ API, and my implementation of it, every day. When it breaks, I see it and live it too. 😄

As I stated in other threads - there's a lot of history in the UA arms race with myQ. Homebridge myQ users are largely fine outside of the load balancer blip late last week - or if they aren't, I'm not hearing about it and the anecdotal traffic on the project Github and on the Homebridge Discord would indicate that as well.

**To reiterate: correlation is not causation. To say that, for some users, adding a UA string allows them to connect to the API when using HA, does not mean that myQ has implemented UA blocking. If that were true, it would apply to every other implementation...it does not.**

I'll add one more postscript: @ehendrix23 you might want to consider upgrading the HA implementation to use HTTP2. I've found pretty substantive performance and reliability improvements. Assuming not much else has changed, it's the one substantive difference I can think of, off the top of my head, between our implementations at this point.

**Comment by lygris** [comment]
> **To reiterate: correlation is not causation. To say that, for some users, adding a UA string allows them to connect to the API when using HA, does not mean that myQ has implemented UA blocking. If that were true, it would apply to every other implementation...it does not.**


You can block things conditionally on a load balancer or firewall, based on many things. They could certainly be blocking UA for http requests and not for http2 as you've just mentioned your code uses. That or the way your plugin is using the API is not matching some criteria for the policy that blocks UA.

It's not like adding a user agent magically fixed the way pymyq is doing the request, so it's fair to say there is something blocking blank user agents when the requests come from pymyq. Hubitat is having the same issue as homeassistant and they don't use pymyq from what i can tell, but specifying a user agent still fixed their issue.

I'm not a coder, I do Azure architecture work. I've put in blocks like this before. Not sure what you're doing different in your code but clearly it's better than pymyq's implementation.

**Comment by starsoccer** [comment]
> I upgraded HA core today and of course myq broke, so I redid the fix and set my UA to "HomeAssistant" and rebooted. received a 403 error on startup. Then went and changed my UA to a well known and widely used UA, rebooted, and myq started working. This is too small of a sample size to fully confirm it's a UA issue, but i think it's fairly telling. Others in HA core issue thread also did state they changed their UA's per my instructions to something like "hass" or other HA related terms/names and reported it did not work. Some however made up a seemingly random string and it did work.
> 
> @starsoccer try this if your request.py is back to the default `docker exec -it homeassistant sed -i.bak "34s/self._useragent = None/self._useragent = str(\"AppleWebKit\/537.36\")/" /usr/local/lib/python3.11/site-packages/pymyq/request.py`

I did try this and still continue to get 403 errors. My errors though are on login not sure if that is what others are seeing as well.

Error:
```
Authentication failed: Error requesting data from https://partner-identity.myq-cloud.com/api/Account/LoginWithEmail?returnUrl=/connect/authorize/callback?client_id%3DIOS_CGI_MYQ%26code_challenge%3XXXXXXXXXX%26code_challenge_method%3DS256%26redirect_uri%3Dcom.myqops%253A%252F%252Fios%26response_type%3Dcode%26scope%3DMyQ_Residential%2520offline_access: 403 - Forbidden
```

**Comment by lygris** [comment]
> > I upgraded HA core today and of course myq broke, so I redid the fix and set my UA to "HomeAssistant" and rebooted. received a 403 error on startup. Then went and changed my UA to a well known and widely used UA, rebooted, and myq started working. This is too small of a sample size to fully confirm it's a UA issue, but i think it's fairly telling. Others in HA core issue thread also did state they changed their UA's per my instructions to something like "hass" or other HA related terms/names and reported it did not work. Some however made up a seemingly random string and it did work.
> > @starsoccer try this if your request.py is back to the default `docker exec -it homeassistant sed -i.bak "34s/self._useragent = None/self._useragent = str(\"AppleWebKit\/537.36\")/" /usr/local/lib/python3.11/site-packages/pymyq/request.py`
> 
> I did try this and still continue to get 403 errors. My errors though are on login not sure if that is what others are seeing as well.
> 
> Error:
> 
> ```
> Authentication failed: Error requesting data from https://partner-identity.myq-cloud.com/api/Account/LoginWithEmail?returnUrl=/connect/authorize/callback?client_id%3DIOS_CGI_MYQ%26code_challenge%3XXXXXXXXXX%26code_challenge_method%3DS256%26redirect_uri%3Dcom.myqops%253A%252F%252Fios%26response_type%3Dcode%26scope%3DMyQ_Residential%2520offline_access: 403 - Forbidden
> ```

`docker exec -it homeassistant sed -n 34p /usr/local/lib/python3.11/site-packages/pymyq/request.py`

Run that and post the result, looks like it didn't change the line

**Comment by starsoccer** [comment]
Seems correct to me
image

**Comment by lygris** [comment]
Hmm, you rebooted HA after the change right? just reloading the integration won't do

**Comment by starsoccer** [comment]
> Hmm, you rebooted HA after the change right? just reloading the integration won't do

Just did another one for sanity sake and no change. Im happy to try to make some requests in postman locally on my machine and see what happens if there is any documentation on how to make the requests

**Comment by DanielBaulig** [comment]
Just chiming in here. I don't really have anything substantive to add to the conversation other than a couple of pieces of guidance that might help shape the conversation and make it move forward.

1) Clearly @hjdhjd has some insider knowledge that he is not able to share directly with us. I think we should give his statements, even if they currently may appear to be contradicting our own experiences, a little more credence. 

2) I see docker execs and sed code changes being thrown around in this thread. While I understand that people are trying to debug this as close as possible to the underlying code, actually being able to reproduce the issue using a different "platform" is going to lend itself more to understanding the true nature of the issue. If we can't reproduce the issue using a simple curl request, then we don't understand the issue sufficiently for a truly solid and robust fix.

3) While changing the UA string might cause some relief in some cases, as some people have clearly experienced, it might very well be that the underlying issue is more complex or nuanced. For example -- and I am just making this up as an example, not necessarily because I believe or even have evidence that this is actually the underlying problem -- rate limiting, load balancing and other forms of network and request management might depend on or include the UA string as a way to identify a specific client. Thus playing more nicely with those mechanisms in the first place might avoid running into any problems regarding the UA string down the line entirely.

**Comment by hjdhjd** [comment]
> Just chiming in here. I don't really have anything substantive to add to the conversation other than a couple of pieces of guidance that might help shape the conversation and make it move forward.
> 
> 1. Clearly @hjdhjd has some insider knowledge that he is not able to share directly with us. I think we should give his statements, even if they currently may appear to be contradicting our own experiences, a little more credence.
> 2. I see docker execs and sed code changes being thrown around in this thread. While I understand that people are trying to debug this as close as possible to the underlying code, actually being able to reproduce the issue using a different "platform" is going to lend itself more to understanding the true nature of the issue. If we can't reproduce the issue using a simple curl request, then we don't understand the issue sufficiently for a truly solid and robust fix.
> 3. While changing the UA string might cause some relief in some cases, as some people have clearly experienced, it might very well be that the underlying issue is more complex or nuanced. For example -- and I am just making this up as an example, not necessarily because I believe or even have evidence that this is actually the underlying problem -- rate limiting, load balancing and other forms of network and request management might depend on or include the UA string as a way to identify a specific client. Thus playing more nicely with those mechanisms in the first place might avoid running into any problems regarding the UA string down the line entirely.

I don’t have inside knowledge beyond being a myQ API elder / grey-hair who’s written the reference implementation of the API that the open source community uses. 😄 And more broadly, I’ve just been at this a really long time.

One thing folks might want to try is to A/B test across the open source ecosystems and see if the problems persist or not. If things break in HASS and in Homebridge, for instance, there could be more going on. By and large, the user community I support (Homebridge directly, SmartThings indirectly) - isn’t running into widespread issues at this point. 

We all had a hiccup due to myQ infrastructure woes last Friday/Saturday. Beyond that, it’s been relative smooth sailing. I’ve taken an additional step to dynamically try across myQ regions when things fail in the most recent API release (yesterday) and that should further add resiliency to reduce issues like the ones we encountered last Friday on the Homebridge end at least. And none of these solutions involve modifying UA…again, this is the open source world, so you folks do you. I’ve made my case why modifying UAs, particularly in larger open source projects like HASS, may have unintended future consequences for all the open source myQ solutions (and a history of myQ/Liftmaster doing so), but it’s your community too. 

Alright, I’m tapping out of these threads…if the devs would like to find me, they know where they can reach me. Best of luck everyone to getting up and running, I hope we can all get back to this being a boring API that “just works” again soon.

**Comment by starsoccer** [comment]
Just to update, about 2 days ago after a reboot myq decided to fix itself for me and worked for a few hours/day. After that I rebooted for something else and it has been broken since. So just to reiterate again, this patch may make the issue less common to occur but it does not 100% fix the ongoing issues.

**Comment by starsoccer** [comment]
Since the other thread has been locked I thought I would post here so atleast anyone else who encounters the same issues as me can hopefully get resolution.

Due to totally unrelated circumstances I was switching phones and needed to login to the myq app today. Upon trying to do I was shown a cloudflare popup saying my request was denied because of bad behavior and an example was listed as being an sql injection attack.

After laughing at how poorly the cloudflare instance was configured, I then did a password reset and changed my password to something containing no special characters and it worked right away. I then repeated this on HA without the useragent fix(it did not work). I then repeated again with the useragent fix applied and it worked.

So for anyone else encountering issues Id first check your password and ensure there are no or very limited special characters present

**Comment by quinat** [comment]
Thanks, changing my password to remove special characters worked for me. I am in Australia.

**Pull request closed without merging**
